rainier-core/src/main/scala/com/stripe/rainier/core/Discrete.scala

```diff
@@ -13,6 +13,78 @@ object Discrete {
     }
 }
 
+/**
+  * Discrete Constant (point mass) with expecation `constant`
+  *
+  * @param constant The integer value of the point mass
+  */
+final case class DiscreteConstant(constant: Real) extends Discrete {
+  val generator: Generator[Int] =
+    Generator.require(Set(constant)) { (c, n) =>
+      n.toInt(constant)
+    }
+
+  def logDensity(v: Real): Real = {
+    If(v - constant, Real.negInfinity, 0)
+  }
+}
+
+/**
+  * Bernoulli distribution with expectation `p`
+  *
+  * @param p The probability of success
+  */
+final case class Bernoulli(p: Real) extends Discrete {
+  val generator: Generator[Int] =
+    Generator.require(Set(p)) { (r, n) =>
+      val u = r.standardUniform
+      val l = n.toDouble(p)
+      if (u <= l) 1 else 0
+    }
+
+  def logDensity(v: Real) =
+    If(v, p.log, (1 - p).log)
+}
+
+/**
+  * Geometric distribution with expectation `1/p`
+  *
+  * @param p The probability of success
+  */
+final case class Geometric(p: Real) extends Discrete {
+  val generator: Generator[Int] =
+    Generator.require(Set(p)) { (r, n) =>
+      val u = r.standardUniform
+      val q = n.toDouble(p)
+      Math.floor(Math.log(u) / Math.log(1 - q)).toInt
+    }
+
+  def logDensity(v: Real) =
+    p.log + v * (1 - p).log
+}
+
+/**
+  * Negative Binomial distribution with expectation `n(1-p)/p`
+  *
+  * @param n Total number of failures
+  * @param p Probability of success
+  */
+final case class NegativeBinomial(n: Real, p: Real) extends Discrete {
+  val generator: Generator[Int] =
+    Generator.require(Set(n, p)) { (r, m) =>
```

Comment:
As you point out, this is linear in n. For large enough n there's a normal approximation; could we switch to that after some threshold?

Reply:
Here's my proposed thresholds for when to switch to a normal distribution: https://gist.github.com/DavidRdgz/bc313a8ef3133a082c7ca10503824e2d

Reply:
:+1: